File: rd-framework/src/test/kotlin/com/jetbrains/rider/framework/test/util/DynamicEntity.kt
package com.jetbrains.rider.framework.test.util

import com.jetbrains.rider.framework.*
import com.jetbrains.rider.framework.base.RdBindableBase
import com.jetbrains.rider.framework.impl.RdProperty
import com.jetbrains.rider.util.lifetime.Lifetime
import com.jetbrains.rider.util.reactive.IProperty
import kotlin.reflect.KClass


class DynamicEntity(val _foo: RdProperty<Boolean?>) : RdBindableBase() {
    val foo: IProperty<Boolean?> = _foo

    companion object : IMarshaller<DynamicEntity> {
        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): DynamicEntity {
            return DynamicEntity(RdProperty.read(ctx, buffer, FrameworkMarshallers.Bool.nullable()))
        }

        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: DynamicEntity) {
            RdProperty.write(ctx, buffer, value._foo)
        }

        override val _type: KClass<*>
            get() = DynamicEntity::class

        fun create(protocol: IProtocol) {
            protocol.serializers.register(DynamicEntity)
        }
    }

    override fun init(lifetime: Lifetime) {
        _foo.bind(lifetime, this, "foo")
    }

    override fun identify(identities: IIdentities, id: RdId) {
        _foo.identify(identities, id.mix("foo"))
    }

    constructor(_foo: Boolean?) : this(RdProperty(_foo, FrameworkMarshallers.Bool.nullable()))
}

File: rd-cpp/src/rd_framework_cpp/src/test/util/entity/DynamicEntity.cpp
#include "DynamicEntity.h"

namespace rd {
	namespace test {
		namespace util {
			DynamicEntity DynamicEntity::read(SerializationCtx  &ctx, Buffer &buffer) {
				return DynamicEntity(RdProperty<int32_t>::read(ctx, buffer));
			}

			std::string DynamicEntity::type_name() const { return "DynamicEntity"; }

			void DynamicEntity::write(SerializationCtx  &ctx, Buffer &buffer) const {
				foo.write(ctx, buffer);
			}

			void DynamicEntity::create(IProtocol *protocol) {
				protocol->serializers->registry<DynamicEntity>();
			}

			void DynamicEntity::init(Lifetime lifetime) const {
				foo.bind(lifetime, this, "foo");
			}

			void DynamicEntity::identify(Identities const &identities, RdId const &id) const {
				foo.identify(identities, id.mix("foo"));
			}

			bool operator==(const DynamicEntity &lhs, const DynamicEntity &rhs) {
				return lhs.foo == rhs.foo;
			}

			bool operator!=(const DynamicEntity &lhs, const DynamicEntity &rhs) {
				return !(rhs == lhs);
			}

			size_t DynamicEntity::hashCode() const noexcept {
				return IPolymorphicSerializable::hashCode();
			}

			bool DynamicEntity::equals(const ISerializable &object) const {
				auto const &other = dynamic_cast<DynamicEntity const&>(object);
				return this == &object;
			}

			std::string DynamicEntity::static_type_name()
			{
				return "DynamicEntity";
			}

			std::string to_string(DynamicEntity const &value) {
				return to_string(value.foo);
			}
		}
	}
}
